### PR TITLE
Create models for database

### PR DESCRIPTION
**Pull request by AndreyKhomich**

### CODE REVIEW

File: backend/database/models.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+from sqlalchemy.ext.declarative import declarative_base
+
+Base = declarative_base()
+
+
+class User(Base):
+    __tablename__ = "user"
+
+    id = Column(Integer, primary_key=True, index=True)
+    name = Column(String(1024))
```
Comment (valeryia-v) [inline backend/database/models.py]: there is no such requirement that User model should have name

File: backend/database/models.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+from sqlalchemy.ext.declarative import declarative_base
+
+Base = declarative_base()
+
+
+class User(Base):
+    __tablename__ = "user"
+
+    id = Column(Integer, primary_key=True, index=True)
```
Comment (valeryia-v) [inline backend/database/models.py]: PKs are indexed by default, so index is not required here

File: backend/database/models.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    name = Column(String(1024))
+    email = Column(EmailType(256))
+    background_color = Column(ColorType)
```
Comment (valeryia-v) [inline backend/database/models.py]: should be just `background`
```suggestion
    background = Column(ColorType)
```

File: backend/database/models.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    user_id = Column(Integer, ForeignKey('user.id'))
+    lesson_id = Column(Integer, ForeignKey('lesson.id'))
+
+
+class Exercise(Base):
+    __tablename__ = "excercise"
+
+    id = Column(Integer, primary_key=True, index=True)
+    lesson_id = Column(Integer, ForeignKey('lesson.id'))
+    type = Column(Enum('option_a', 'option_b', 'option_c', 'option_d', name='answer_types'))
```
Comment (valeryia-v) [inline backend/database/models.py]: 1. can declare Enum class with all possible options, to reuse in the future. Can also be passed : https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.Enum. In this case you also won't need to specify the name.
2. options should be separate DB columns as per the task. Exercise type should be one of: multiple-choice, free.

File: backend/database/models.py (hunk 2 above)
Comment (valeryia-v) [reply backend/database/models.py]: also need `autoincrement=True` here

Comment (valeryia-v) [reply backend/database/models.py]: same for all the other ids

File: backend/database/models.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    streak = Column(Integer)
+
+class Lesson(Base):
+    __tablename__ = "lesson"
+
+    id = Column(Integer, primary_key=True, index=True)
+    name = Column(String(1024))
+    content = Column(String)
+    order = Column('0', unique=True)
```
Comment (valeryia-v) [inline backend/database/models.py]: order should be an integer

File: backend/database/models.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    id = Column(Integer, primary_key=True, index=True)
+    user_id = Column(Integer, ForeignKey('user.id'))
+    lesson_id = Column(Integer, ForeignKey('lesson.id'))
+
+
+class Exercise(Base):
+    __tablename__ = "excercise"
+
+    id = Column(Integer, primary_key=True, index=True)
+    lesson_id = Column(Integer, ForeignKey('lesson.id'))
```
Comment (valeryia-v) [inline backend/database/models.py]:
```suggestion
    lesson_id = Column(Integer, ForeignKey(Lesson.id))
```

File: backend/database/models.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    content = Column(String)
+    order = Column('0', unique=True)
+    preview = Column(String)
+
+
+class Progress(Base):
+    __tablename__ = "progress"
+
+    id = Column(Integer, primary_key=True, index=True)
+    user_id = Column(Integer, ForeignKey('user.id'))
```
Comment (valeryia-v) [inline backend/database/models.py]:
```suggestion
    user_id = Column(Integer, ForeignKey(User.id))
```

File: backend/database/models.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    lesson_id = Column(Integer, ForeignKey('lesson.id'))
```
Comment (valeryia-v) [inline backend/database/models.py]:
```suggestion
    lesson_id = Column(Integer, ForeignKey(Lesson.id))
```

File: backend/database/models.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+class Exercise(Base):
+    __tablename__ = "excercise"
+
+    id = Column(Integer, primary_key=True, index=True)
+    lesson_id = Column(Integer, ForeignKey('lesson.id'))
+    type = Column(Enum('option_a', 'option_b', 'option_c', 'option_d', name='answer_types'))
+    definition = Column(String(512))
+    text = Column(String(2048))
+    link = Column(String(512), optional=True)
```
Comment (valeryia-v) [inline backend/database/models.py]: Column has no such parameter as `optional`. Please refer the official documentation: https://docs.sqlalchemy.org/en/14/core/metadata.html

File: backend/database/models.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    answer = Column(String(2048))
```
Comment (valeryia-v) [inline backend/database/models.py]: need to add a constraint to ensure this field has a value only if type=free

File: backend/database/models.py (hunk 7 above)
Comment (valeryia-v) [inline backend/database/models.py]: also need to cascade delete all progress records if a user is deleted

**Review by valeryia-v — CHANGES_REQUESTED** [review]

File: backend/database/models.py (hunk 4 above)
Comment (AndreyKhomich) [reply backend/database/models.py]: Don't understand how to mark columns option_a, option_b, option_c, option_d and answers with exercise type value.

```
class ExerciseType(enum.Enum):
    FREE = "free"
    MULTIPLE_CHOICE = "multiple-choice"

class Exercise(Base):
    __tablename__ = "excercise"
    __table_args__ = (
        CheckConstraint('answers'),
    )
```

    id = Column(Integer, primary_key=True, autoincrement=True)
    lesson_id = Column(Integer, ForeignKey('Lesson.id'))
    type = Column(Enum(ExerciseType))
    definition = Column(String(512))
    text = Column(String(2048))
    link = Column(String(512))
    option_a = Column(String)
    option_b = Column(String)
    option_c = Column(String)
    option_d = Column(String)
    answers = Column(String(2048))

**Pull request closed without merging**